src/main/java/bio/terra/service/iam/sam/SamIam.java

```diff
@@ -250,12 +252,35 @@ private String syncOnePolicy(AuthenticatedUserRequest userReq,
                                                     IamResourceType iamResourceType,
                                                     UUID resourceId) throws ApiException {
         ResourcesApi samResourceApi = samResourcesApi(userReq.getRequiredToken());
-        List<AccessPolicyResponseEntry> results =
-            samResourceApi.listResourcePolicies(iamResourceType.toString(), resourceId.toString());
-        return results.stream().map(entry -> new PolicyModel()
-            .name(entry.getPolicyName())
-            .members(entry.getPolicy().getMemberEmails()))
-            .collect(Collectors.toList());
+        try (Stream<AccessPolicyResponseEntry> resultStream =
+                 samResourceApi.listResourcePolicies(iamResourceType.toString(), resourceId.toString()).stream()) {
+            return resultStream.map(entry -> new PolicyModel()
+                .name(entry.getPolicyName())
+                .members(entry.getPolicy().getMemberEmails()))
+                .collect(Collectors.toList());
+        }
+    }
+
+    @Override
+    public Map<IamRole, String> retrievePolicyEmails(AuthenticatedUserRequest userReq,
+                                                     IamResourceType iamResourceType,
+                                                     UUID resourceId) throws InterruptedException {
+        SamRetry samRetry = new SamRetry(configurationService);
+        return samRetry.perform(() -> retrievePolicyEmailsInner(userReq, iamResourceType, resourceId));
+    }
+
+    private Map<IamRole, String> retrievePolicyEmailsInner(AuthenticatedUserRequest userReq,
```

Comment:
Unfortunately not really.  It's admittedly similar but this was explicitly to pull out policy emails.  The other method pulls out a PolicyModel which returns the SAM principals that have access instead of the lower level policy holders

Reply:
It is a difference in structure, not content. You are making a flat list of principals. The other make a list of policy types with the list of principals inside.

I'm not arguing what list you need. Just wondering why you put the specialization here instead of on the caller side using the results of retrievePolicies.

Reply:
It's all coming from the SAM `AccessPolicyResponseEntry`.  The existing method was interested in returning `policyName` and `policy.memberEmails`.  This new method is specifically to return `policyName` and `email` (which is a different value...this is the `policy-<uuid>@dev.test.firecloud.org` email which is what is used under the covers to control ACLs in Google.  I could have maybe added that to our PolicyModel but I wasn't sure if it made sense to change that object

Reply:
Oh maybe to your point, the only shared bit was the call to getPolicies but the mapping was the main difference

Reply:
Oh, duh. My mistake. Nevermind